File: day23/coprocessor_conflag.pl
#!/usr/bin/env perl
use strict;
use warnings;
use feature qw(:5.24);
use experimental qw(signatures);

my %r = map { $_ => 0 } 'a'..'h';
$r{mul} = 0;
my @pgm;

# parse the input into an array of closures
while (<>) {
    chomp;
    my @tok = split / /;
    if ($tok[0] eq 'set') {
        push @pgm, sub ($ip, $r) {
            my $val = $tok[2] =~ /\d+/ ? $tok[2] : $r->{$tok[2]};
            $r->{$tok[1]} = $val;
            return $ip + 1;
        }
    } elsif ($tok[0] eq 'sub') {
        push @pgm, sub ($ip, $r) {
            my $val = $tok[2] =~ /\d+/ ? $tok[2] : $r->{$tok[2]};
            $r->{$tok[1]} -= $val;
            return $ip + 1;
        }
    } elsif ($tok[0] eq 'mul') {
        push @pgm, sub ($ip, $r) {
            my $val = $tok[2] =~ /\d+/ ? $tok[2] : $r->{$tok[2]};
            $r->{$tok[1]} *= $val;
            $r->{mul}++;
            return $ip + 1;
        }
    } elsif ($tok[0] eq 'jnz') {
        push @pgm, sub ($ip, $r) {
            my $val1 = $tok[1] =~ /([a-z])/ ? $r->{$tok[1]} : $tok[1];
            my $val2 = $tok[2] =~ /([a-z])/ ? $r->{$tok[2]} : $tok[2];
            if ($val1 != 0) {
                return $ip + $val2;
            } else {
                return $ip+1;
            }
        }
    } else {
        say "WTF? $_";
    }
}

# run the program
my $ip = 0;
my $cnt = 0;
$r{a} = 0;
while ($ip >= 0 && $ip < @pgm) {
    $ip = $pgm[$ip]->($ip, \%r);
    $cnt++;
}

say "result1: $r{mul}";

say "$cnt steps";
say "registers:";
say "$_, $r{$_}" for 'a'..'h';

File: day23/coprocessor_conflag.py
#!/usr/bin/env python3
from sys import argv
import re

class Set:
    def __init__(self, tok1, tok2):
        self.tok1 = tok1
        self.tok2 = tok2

    def run(self, ip, r):
        val = int(self.tok2) if re.match('\d+', self.tok2) else r[self.tok2]
        r[self.tok1] = val
        return ip+1

class Sub:
    def __init__(self, tok1, tok2):
        self.tok1 = tok1
        self.tok2 = tok2

    def run(self, ip, r):
        val = int(self.tok2) if re.search('\d+', self.tok2) else int(r[self.tok2])
        r[self.tok1] -= val
        return ip+1

class Mul:
    def __init__(self, tok1, tok2):
        self.tok1 = tok1
        self.tok2 = tok2

    def run(self, ip, r):
        val = int(self.tok2) if re.search('\d+', self.tok2) else r[self.tok2]
        r[self.tok1] *= val
        r['mul'] += 1
        return ip+1

class Jnz:
    def __init__(self, tok1, tok2):
        self.tok1 = tok1
        self.tok2 = tok2

    def run(self, ip, r):
        val1 = r[self.tok1] if re.match('[a-z]', self.tok1) else int(self.tok1)
        val2 = r[self.tok2] if re.match('[a-z]', self.tok2) else int(self.tok2)
        if val1 != 0:
            return ip + val2
        else:
            return ip+1

r = { chr(i): 0 for i in range(ord('a'), ord('h')+1) }
r['mul'] = 0

# parse the input into an array of object
pgm = []
filename = argv[1];
with open(filename) as f:
    for line in f:
        line = line.rstrip()
        tok = line.split(' ');
        if tok[0] == 'set':
            pgm.append(Set(tok[1], tok[2]))
        elif tok[0] == 'sub':
            pgm.append(Sub(tok[1], tok[2]))
        elif tok[0] == 'mul':
            pgm.append(Mul(tok[1], tok[2]))
        elif tok[0] == 'jnz':
            pgm.append(Jnz(tok[1], tok[2]))
        else:
            print("WTF?", line)

# run the program
ip = 0
while True:
    try:
        ip = pgm[ip].run(ip, r)
    except IndexError:
        print("result1:", r['mul'])
        break
